frontend/lib/resolution/VarScopeVisitor.cpp

```diff
@@ -519,6 +571,53 @@ void VarScopeVisitor::exit(const Conditional* cond, RV& rv) {
   exitAst(cond);
 }
 
+bool VarScopeVisitor::enter(const Select* sel, RV& rv) {
+  enterAst(sel);
+  enterScope(sel, rv);
+  
+  //traverse the not paramFalse children
+  for(int i = 0; i < sel->numWhenStmts(); i++) {
+    auto whenAst = sel->whenStmt(i);
+    bool anyParamTrue = false;
+    bool allParamFalse = !whenAst->isOtherwise();
+    for(auto caseExpr : whenAst->caseExprs()) {
+      auto res = rv.byAst(caseExpr);
+      anyParamTrue = anyParamTrue || res.type().isParamTrue();
+      allParamFalse = allParamFalse && res.type().isParamFalse();
+    }
+    if (!allParamFalse) {
+      whenAst->traverse(rv);
+      currentWhenFrame(i)->paramTrue = anyParamTrue;
```

Comment:
I think that in this loop, `anyParamTrue` and `allParamFalse` refer to the list of cases for each `when` statement rather than all the `when` statements together. 

So for the code example above, for `when b`, the `caseExpr` loop calculates `anyParamTrue=false` and `allParamFalse=false`. Because `allParamFalse` is false, it traverses the body of the when node.

Then for `when true`, the `caseExpr` loop calculates `anyParamTrue=true` and `allParamFalse=false`. As above it traverses the body of the when node, but this time sets the frame's `paramTrue` field to true. Then, because `anyParamTrue` is true, breaks from the loop over the `when` nodes (here with no effect bc its the last `when`, but this would matter in other cases).

Reply:
I missed a key piece above. As far as I'm aware, there isn't a "param-selected" mechanism that triggers when a `when` condition is paramTrue at this point. Rather, when a `when` is paramTrue, it just stops any subsequent `when`s from being processed.